### PR TITLE
Update active region parametrization to be smooth + large re-work of code front-end and notebooks

### LINKED ISSUE #5: Small website
Build a small website to describe the installation and implementation of the code. Add there the comparison of Sajax, fleck and spotter both in terms of the mismatch between then (hopefully not) over the parameter space of possible active region configurations, as well as a speed comparison with a basic MCMC retrieval.

### LINKED ISSUE #13: Gradient flow
There is still a clear issue with the flow of gradients through the model. We either need to figure out how to update the masking procedure to allow for gradient to flow easily, or if the masking approaching becomes too complicated / troublesome we need to re-think how we're implementing active regions / planets in the model

### LINKED ISSUE #15: Add an option / toggle to setup a umbra+penubra active region

### PR DESCRIPTION
**Pull request by SamMerc**
## Updated active region modelling
- Active regions (AR) are described as contrast surfaces — each AR is its own super-Gaussian "bump", so overlapping regions compound instead of one hard-masking the other. This also replaces the hard/soft mask distinction used prior commit. An additional parameter is introduced: the smoothness, which serves to move ARs' contrast curve from Gaussian to top-hat functions.
- Each AR can carry its own spectrum and limb-darkening coefficients, not just a shared one. Rotational broadening moved to the spectral level — each pixel's local spectrum is Doppler-shifted before its contrast is read off.

## Updated notebooks
- quickstart.ipynb was rebuilt around real PHOENIX spectra + ExoTic-LD coefficients (replacing static .txt fixtures), restructured into six worked, animated examples.
- inference.ipynb was filled in to show an example fitting both a white light curve and chromatic light curves, with both a gradient descent and an MCMC approach
-  comparison.ipynb was updated with the new code structure, and fleck/spotter transit codes were added

## General code updates
- Merged build_model/build_combined_model into build_system, and evaluate_light_curve/compute_light_curve/compute_combined_light_curve into make_lc/quick_lc.
- build_system takes times/P_rot directly; transit params are an optional all-or-nothing group.
- make_lc supports dynamic (traced) overrides of transit geometry and quiet LDC coefficients per call.
- Reintroduced a soft planet mask — but this time opt-in via transit_softness (default off, so the accuracy fix from commit 1 stays intact for the default hard-edge path), specifically to let jax.grad see transit-geometry parameters for gradient-based retrieval.
- Adds chromatic transit depth (k as a per-wavelength array).
- Drops the epsilon output; returns (lc, star_maps) tuple.
- Renames LdcMode/ldc_* → LdMode/ld_*.

This PR directly addresses issues #5 , #13 , and #15

### CODE REVIEW

File: sajax/core.py [hunk 1]
```diff
@@ -12,59 +12,93 @@
...
+*Do NOT jit(make_lc) directly* -- it contains Python-level
 control flow on model metadata.  Instead, the inner _compute_all_phases
 is the hot path and is safe to JIT via:
 
     from jax import jit
     _compute_all_phases_jit = jit(_compute_all_phases, static_argnames=[
-        "star_pixel_rad", "total_pixels", "ldc_mode",
-        "ar_overlap_mode", "plot_map_wavelength", "n",
+        "star_pixel_rad", "total_pixels", "ld_mode",
+        "plot_map_wavelength", "n",
```
Comment (lmarthur) [inline sajax/core.py]:
```suggestion
        "plot_map_wavelength", "n", "transit_softness"
```

File: sajax/core.py [hunk 2]
```diff
@@ -257,420 +330,612 @@ def _compute_ar_mask(
...
-            Specific intensity at each (wavelength, μ) grid point.
-            Required when ``ldc_mode="intensity_profile"``.
-    phases_rot : array_like, shape (nphase,)
+        Quiet-photosphere flux / spectrum.
+    times : array_like, shape (ntime,)
+        Absolute observation times [days].
+    P_rot : float
+        Stellar rotation period [days].
     stellar_grid_size : int
+        Size of the stellar grid (number of pixels along one side of the square grid).
```
Comment (lmarthur) [inline sajax/core.py]:
```suggestion
        Radius of the stellar grid.
```

File: sajax/core.py [hunk 3]
```diff
@@ -897,307 +1336,289 @@ def _rotate_ars_at_phase(phase_deg):
...
+        ``flux_active``/``ar_lat``/``ar_long``/``ar_size``/``ar_smoothness``
+        are all-or-nothing: give every one of them to add active region(s),
+        or omit all five for a quiet star. Giving some but not all raises
+        ``ValueError``.
+    times : array_like, shape (ntime,)
+        Absolute observation times [days].
+    P_rot : float
+        Stellar rotation period [days].
     stellar_grid_size : int
+        Size of the stellar grid (number of pixels along one side of the square grid).
```
Comment (lmarthur) [inline sajax/core.py]:
```suggestion
        Radius of the stellar grid.
```

File: docs/usage.md [hunk 4]
```diff
@@ -23,7 +23,7 @@ It takes stellar spectra, active region properties, and rotation parameters to p
 | `ar_size` | Angular radius of active region [deg] | `[10.0]` |
 | `phases_rot` | Rotation phases [deg] | `np.linspace(0, 360, 50)` |
```
Comment (lmarthur) [inline docs/usage.md]:
```suggestion
| `ar_smoothness` | Super-Gaussian order of the AR edge (higher = sharper) | `[4.0]` |
| `times` | Absolute observation times [days] | `np.linspace(0, 10, 50, endpoint=False)` |
| `P_rot` | Stellar rotation period [days] | `10.0` |
```

File: docs/usage.md
Comment (lmarthur) [inline docs/usage.md]: Epsilon is no longer a supported parameter, correct? This doc section will need to be updated to match the new implementation.

File: docs/usage.md [hunk 5]
```diff
@@ -76,7 +76,7 @@ params = dict(
     inc_star    = 90.0,     # stellar inclination [deg]  (equator-on)
 )
 
-result = compute_light_curve(
+result = quick_lc(
```
Comment (lmarthur) [inline docs/usage.md]: The function signature should be updated to match the new arguments.

File: sajax/core.py [hunk 6]
```diff
@@ -232,23 +292,36 @@ def _make_oversampled_phases(
...
+    Super-Gaussian shape over in-disc pixels, centred on the active region.
 
-    Because ``x_disc`` and ``y_disc`` already contain only stellar-disc
-    pixels, there is no need for an in-star check here.
+    The AR is a spherical cap of angular radius ``arsize_rad`` (the distribution's
+    "sigma"). The shape falls off from the AR's centre as a super-Gaussian
+    of order ``ar_smoothness``:
+
+      - ``ar_smoothness -> inf``  converges to a top-hat function.
+      - ``ar_smoothness == 1``    is a Gaussian.
```
Comment (lmarthur) [inline sajax/core.py]: I want to make sure that I am following this correctly. 

Is the ar_smoothness == 1 case intended to be Gaussian in theta? Or in x = 1 - cos(theta)?

File: tests/test_core.py
Comment (lmarthur) [inline tests/test_core.py:1]: I may be missing something, but I'm not clear on the intended behavior for ar_smoothness <= 0.

For ar_smoothness = 0, the exponent is 0, leading to a constant exp(-1) value everywhere, which corresponds to a uniform dimming of the star. 

For ar_smoothness < 0, does the active region become a facula?

We should make sure that the behavior matches what we want, which may mean restricting ar_amoothness > 0, and these edge cases should be added to the test suite to enforce the correct handling.

File: sajax/core.py [hunk 7]
```diff
@@ -897,307 +1336,289 @@ def _rotate_ars_at_phase(phase_deg):
...
+    # ---- Single-wavelength convenience: drop the now-degenerate nwave
+    # axis so single-channel callers get (nphase,) instead of (nphase, 1) ---
+    if nwave == 1:
+        lc_raw = lc_raw[..., 0]
 
+    return lc_raw, star_maps
 
-def compute_light_curve(
+
+def quick_lc(
```
Comment (lmarthur) [inline sajax/core.py:1457]: Should transit_softness be an argument of quick_lc?

**Comment by lmarthur** [comment]
I've left a handful of small, specific comments and proposed edits. Unfortunately it's not easy to directly propose edits to diffs for .ipynb files, so I'm including a few specific items below. Most of these are small consistency checks, but a few of them are larger conceptual things that I want to make sure I understand.

## docs/comparison.ipynb
- In the Transit Light Curve section, how does spotter get the a_over_rstar value? It doesn't look like it's passed directly. I'm not familiar enough with spotter to know how it works internally, but I wonder if the relatively large spotter residuals could be a consequence of a slightly different orbital geometry. 
- In the third code cell, the latitude and longitude comments are swapped. 
- In Section 3, `axes['D'].set_title()` is called twice so the fleck residuals are untitled but the spotter title is set twice. 
- There are two sections labeled 'Section 3'.

## docs/inference.ipynb
- The white lightcurve is built by averaging 10 different curves. However, these curves appear to be independently normalized before they are combined, but they should not contribute evenly to the white lightcurve. If I understand this correctly, this probably leads to a bias in the recovery. 
- I think the second `sampler_white.print_summary()` should be `sampler_2d.print_summary()` (in the Summary and Traces section).
- Cells 15 and 18 refer to an array with 100 wavelengths, but NWAVE is set to 10 in cell 4.
- In cell 10, `period`, `inclination`, etc. are marked as placeholders, but `k` is also overridden and is not marked accordingly.
- T14_approx in the Ground Truth section uses hardcoded values, which could drift from the above variables. It also uses 5.0 for the period, which is the *rotational* period. I think it should be 4.2, the *orbital* period.

## docs/quickstart.ipynb
- In Case 3, the umbra_flux and penumbra_flux are specified explicitly, but I think the true flux at the center of the AR is generated by compounding both effects. I don't have a clear intuition for why this central flux should be (1 - 0.15 - 0.65 = 0.2) as opposed to (1 * 0.85 * 0.35 = 0.2975), or even just 0.35 itself (no compounding effect). This is a broader comment than just for this notebook, but this is where the question came up for me. 
- In cell 11, the super-Gaussian plots `shape = jnp.exp(-x / x0) ** n)` which does not match the code. It should be `shape = jnp.exp(-x / x0) ** (2 * n))`. I think this is why the n=1 case looks slightly off. 
- In cell 6, I suspect that `Wavelength range: 6000 - 25000 nm` has the incorrect unit labels. Units should either be re-labeled as Angstroms, or the numerical values should be divided by 10 to convert to nm. 
- The case numbering has minor inconsistencies. The "Case 4" heading uses `result_case5`, the "Case 5" heading uses `lc_combined6` and `_update_case6`, and the plot title reads "Case 6". 
- The `params_case4` dict is defined but never referenced.
- Cell 11 says `# (ntimes, 100, 100)` for the shape of the star map with `stellar_grid_size = 100`, but `stellar_grid_size` is a radius, not a side length, so the actual shape is `(ntimes, 201, 201)`.

**Review by lmarthur — CHANGES_REQUESTED** [review]

File: docs/usage.md
Comment (SamMerc) [reply docs/usage.md]: Good call, I'm removing epsilon since it's no longer a supported parameter.

File: tests/test_core.py
Comment (SamMerc) [reply tests/test_core.py:1]: Good point. The ar_smoothness is not supposed to be <1, otherwise it leads to weird behaviour, so I'll enforce that in the code (throw an error if it is <1) and will add some test cases for it.

File: docs/usage.md
Comment (SamMerc) [reply docs/usage.md]: This has been dealt with in commit [ca91a6a](https://github.com/SamMerc/sajax/pull/17/commits/ca91a6a03f4676ffb7b9567b9d1ba71a703a629f)

File: docs/usage.md (hunk 5 above)
Comment (SamMerc) [reply docs/usage.md]: This has been dealt with in commit [ca91a6a](https://github.com/SamMerc/sajax/pull/17/commits/ca91a6a03f4676ffb7b9567b9d1ba71a703a629f)

File: sajax/core.py (hunk 6 above)
Comment (SamMerc) [reply sajax/core.py]: Good catch, this was a mistake in the code on my part. 

`ar_smoothness` is a super-Gaussian in x and not in theta itself. x is just an auxiliary variable, it has no relation to a physical distance. We use x instead of theta directly because computing theta requires an arccos call which has an infinite derivative at theta=0 causing a jax.grad error later down the line. But in the end what we really care about is what the super-Gaussian looks like as a function of theta, because that's what our active region's size is defined as (an angle that is).

In this context, there was a bug in the way the super-Gaussian was computed. Since x~0.5*theta^2 for small angles, the old exponent `2*ar_smoothness` effectively squared this at small angles giving a theta^4 function near the super-Gaussian core which results in a more flat-topped curve than a real Gaussian. 

The solution is therefore to drop the factor of 2 in the exponent so that `ar_smoothness = 1` corresponds to the truly Gaussian case. (check out the quickstart notebook, the n=1 case now looks like a true Gaussian on the stellar surface).

In fixing this bug, I also updated the precision of the super-Gaussian function calculation: x and x0 were computed as 1-cos(theta), which cancels to 0 in float32 for small angular radii, so using the formula chord length directly is safer.

In total, I updated core.py, two test cases related to `ar_smoothness` in test_core.py and all the notebooks to have updated plots.

All these updates are in commit [b677203](https://github.com/SamMerc/sajax/pull/17/commits/b677203facc0b1f19ef477d0d16cc840290cdf79)

File: tests/test_core.py
Comment (SamMerc) [reply tests/test_core.py:1]: This has been dealt with in commits [b9c4140](https://github.com/SamMerc/sajax/pull/17/commits/b9c41409d218e2375a651a3297e1d299757110fb) and [15a3bf3](https://github.com/SamMerc/sajax/pull/17/commits/15a3bf32e2d06a9de98cc2daa436e61618c7d6c5)

File: sajax/core.py (hunk 7 above)
Comment (SamMerc) [reply sajax/core.py:1457]: transit_softness only matters in the context of fitting where gradients would be required so ideally we would only pass it to make_lc and not quick_lc. 
An argument can be made about keeping it as a possible option to quick_lc but I think it would be better to keep quick_lc and make_lc separate so as to not muddy the water about which one should be used in which context.

**Comment by SamMerc** [comment]
Just finished responding to your code suggestions and comments. Here are my answers to your questions about the notebooks.

## docs/comparison.ipynb
- In the Transit Light Curve section, how does spotter get the a_over_rstar value? It doesn't look like it's passed directly. I'm not familiar enough with spotter to know how it works internally, but I wonder if the relatively large spotter residuals could be a consequence of a slightly different orbital geometry. 
  -  spotter takes in a/R* through the impact parameter passed to the `Body` class.
- In the third code cell, the latitude and longitude comments are swapped. 
  - Good catch, this has been dealt with in commit [b81dda9](https://github.com/SamMerc/sajax/pull/17/commits/b81dda959356674a110711873096f698e3792f3d)
- In Section 3, `axes['D'].set_title()` is called twice so the fleck residuals are untitled but the spotter title is set twice. 
  - This has been dealt with in commit [9b026ed](https://github.com/SamMerc/sajax/pull/17/commits/9b026ed6ffa9a893c1c0e7891c745815ed226623)
- There are two sections labeled 'Section 3'.
  - This has been dealt with in commit [9b026ed](https://github.com/SamMerc/sajax/pull/17/commits/9b026ed6ffa9a893c1c0e7891c745815ed226623)

## docs/inference.ipynb
- The white lightcurve is built by averaging 10 different curves. However, these curves appear to be independently normalized before they are combined, but they should not contribute evenly to the white lightcurve. If I understand this correctly, this probably leads to a bias in the recovery. 
  - Thanks for finding that, yes indeed the white light curve was not being built correctly. It is now averaged over the un-normalized chromatic light curves before being normalized on its own. This was included in commit [8cf320f](https://github.com/SamMerc/sajax/pull/17/commits/8cf320f3d086f3af8983c1582e220e131e3823cf)
- I think the second `sampler_white.print_summary()` should be `sampler_2d.print_summary()` (in the Summary and Traces section).
  - Yes, indeed, thanks for catching that mistake. This has been updated in commit [8cf320f](https://github.com/SamMerc/sajax/pull/17/commits/8cf320f3d086f3af8983c1582e220e131e3823cf)
- Cells 15 and 18 refer to an array with 100 wavelengths, but NWAVE is set to 10 in cell 4.
  - This has been dealt with in commit [8cf320f](https://github.com/SamMerc/sajax/pull/17/commits/8cf320f3d086f3af8983c1582e220e131e3823cf)
- In cell 10, `period`, `inclination`, etc. are marked as placeholders, but `k` is also overridden and is not marked accordingly.
  - This has been dealt with in commit [8cf320f](https://github.com/SamMerc/sajax/pull/17/commits/8cf320f3d086f3af8983c1582e220e131e3823cf)
- T14_approx in the Ground Truth section uses hardcoded values, which could drift from the above variables. It also uses 5.0 for the period, which is the *rotational* period. I think it should be 4.2, the *orbital* period.
  - Thanks for pointing that out. I swapped out the approximate calculation for a true calculation with no hardcoded values, just using the variables defined. I swapped this out in quickstart notebook as well, which utilized the same T14_approx calculation. This was all done in commit [8cf320f](https://github.com/SamMerc/sajax/pull/17/commits/8cf320f3d086f3af8983c1582e220e131e3823cf)

## docs/quickstart.ipynb
- In Case 3, the umbra_flux and penumbra_flux are specified explicitly, but I think the true flux at the center of the AR is generated by compounding both effects. I don't have a clear intuition for why this central flux should be (1 - 0.15 - 0.65 = 0.2) as opposed to (1 * 0.85 * 0.35 = 0.2975), or even just 0.35 itself (no compounding effect). This is a broader comment than just for this notebook, but this is where the question came up for me. 
  - Owing to the fact that the active regions are described as overlapping contrast curves, their effects are additive. Therefore, as you correctly wrote, the results of the central flux as seen on the corresponding stellar surface map is (1 - (1-0.85) - (1 - 0.35)) = (1 - 0.15 - 0.65) = 0.2. (1 * 0.85 * 0.35 = 0.2975) would mean that their contrast multiply which is not the case, and 0.35 would be if we had a hard, top-hat mask as we had previously. The issue here is likely related to the super-gaussian bug that I mention in an earlier comment. Let me know what you think of the new plots, hopefully they look better.
- In cell 11, the super-Gaussian plots `shape = jnp.exp(-x / x0) ** n)` which does not match the code. It should be `shape = jnp.exp(-x / x0) ** (2 * n))`. I think this is why the n=1 case looks slightly off. 
  - This plot has been updated and the n=1 case should now look like an actual Gaussian. See comments higher up for an extensive discussion on this.
- In cell 6, I suspect that `Wavelength range: 6000 - 25000 nm` has the incorrect unit labels. Units should either be re-labeled as Angstroms, or the numerical values should be divided by 10 to convert to nm. 
  - Thank you, yes, this has been re-labeled to Angstroms [1c91442](https://github.com/SamMerc/sajax/pull/17/commits/1c914423c4b5b47b9b4d71ca8efe8d2146f0c30d)
- The case numbering has minor inconsistencies. The "Case 4" heading uses `result_case5`, the "Case 5" heading uses `lc_combined6` and `_update_case6`, and the plot title reads "Case 6". 
  - Thanks for finding these little inconsistencies, they have been dealt with in commit [1c91442](https://github.com/SamMerc/sajax/pull/17/commits/1c914423c4b5b47b9b4d71ca8efe8d2146f0c30d)
- The `params_case4` dict is defined but never referenced.
  - This has been removed in commit [1c91442](https://github.com/SamMerc/sajax/pull/17/commits/1c914423c4b5b47b9b4d71ca8efe8d2146f0c30d)
- Cell 11 says `# (ntimes, 100, 100)` for the shape of the star map with `stellar_grid_size = 100`, but `stellar_grid_size` is a radius, not a side length, so the actual shape is `(ntimes, 201, 201)`.
  - Dropped this entirely because on top of the side length being 201 instead of 100, a mask is applied to ensure we only keep the grid cells within the stellar radius so the x/y positional arrays have varying lengths depending on longitude/latitude considered. Dealt with in commit [1c91442](https://github.com/SamMerc/sajax/pull/17/commits/1c914423c4b5b47b9b4d71ca8efe8d2146f0c30d)

**Review by lmarthur — APPROVED** [review]
These changes look good! I think this is ready to merge.
